lisa/_assets/kmodules/sched_tp/perf_counters.c

```diff
@@ -0,0 +1,694 @@
+};
+
+#define PERFCTR_DESC(__name, __id, __type, __en)				\
+	((struct perfctr_desc) {						\
+		.name = __name, .id = __id, .type = __type, .default_on = __en,	\
+	})
+
+#define PERFCTR_DESC_COUNT_HW(__name, __id, __en)	\
+	PERFCTR_DESC(__name, __id, PERF_TYPE_HARDWARE, __en)
+
+/* To be extended if needed */
+static const struct perfctr_desc perfctr_generic_lt [] = {
+	PERFCTR_DESC_COUNT_HW("cpu_cycles", PERF_COUNT_HW_CPU_CYCLES, 1),
+};
+
+#define PERF_COUNTERS_PARAM(name, desc)				\
+	static char *name[MAX_PERF_COUNTERS];			\
+	static unsigned int name##_count;			\
+	module_param_array(name, charp, &name##_count, 0644);	\
+	MODULE_PARM_DESC(name, desc);
+
+PERF_COUNTERS_PARAM(perfctr_events_param, "counters");
+
+
+struct perfctr_event_entry {
+	struct hlist_node		node;
+	struct hlist_node		group_link;
+	struct perf_event		*event;
+	struct perfctr_event_group	*group;
+};
+
+struct perfctr_event_group {
+	struct list_head	node;
+	struct hlist_head	entries;
+	unsigned int		raw_id;
+};
+
+struct perfctr_pcpu_data {
+	struct hlist_head	events;
+	struct work_item	*work;
+};
+
+struct perfctr_core {
+	struct list_head			events;
+	struct mutex				lock;
+	struct perfctr_pcpu_data __percpu	*pcpu_data;
+	unsigned int				nr_events;
+	unsigned int				max_nr_events;
+	unsigned long				sampling_period;
+	struct kobject				kobj;
+	struct completion			kobj_unregister;
+	struct srcu_struct			srcu;
+};
+
+static int perfctr_worker_fn(void *data)
+{
+	struct perfctr_core *perf_data = data;
+	struct perfctr_event_entry *entry;
+	struct hlist_head *entry_list;
+	int cpu = smp_processor_id();
+	u64 value, enabled, running;
+	int id;
+
+	if (!trace_perf_counter_enabled())
+		goto leave;
+
+	entry_list = &per_cpu_ptr(perf_data->pcpu_data, cpu)->events;
+
+	if (hlist_empty(entry_list))
+		/* Re-enabled once new events get activated */
+		return WORKER_DISABLE;
+
+	id = srcu_read_lock(&perf_data->srcu);
+	hlist_for_each_entry_srcu(entry, entry_list, node,
+				  srcu_read_lock_held(&perf_data->srcu)) {
+		value = perf_event_read_value(entry->event,
+					      &enabled,
+					      &running);
+		trace_perf_counter(cpu, entry->group->raw_id, value);
+	}
+	srcu_read_unlock(&perf_data->srcu, id);
+leave:
+	return perf_data->sampling_period;
+}
+
+static __maybe_unused
+void perfctr_reactivate_workers(struct perfctr_core *perf_data)
+{
+	int cpu;
+
+	if (!perf_data->nr_events)
+		return;
+
+	for_each_online_cpu(cpu) {
+		 struct work_item *work;
+
+		 work = per_cpu_ptr(perf_data->pcpu_data, cpu)->work;
+		 if (WARN_ON(!work)) /* hot-plug */
+		 	 continue;
+		 restart_work(work, perf_data->sampling_period);
+	}
+}
+
+static void perfctr_event_release_entry(struct perfctr_event_entry *entry);
+
+static int perfctr_event_activate_single(struct perfctr_core *perf_data,
+					 struct perf_event_attr *attr)
+{
+	struct perfctr_event_group *group;
+	struct perfctr_event_entry *entry;
+	struct hlist_node *next;
+	int cpu;
+
+	WARN_ON(!mutex_is_locked(&perf_data->lock));
```

Comment:
is WARN_ON appropriate ? If we are supposed to take that lock, we really should never mess with anything it covers if the lock is not taken, not just merely warn about it.